pkg/operator/rotate.go

```diff
@@ -0,0 +1,242 @@
+package operator
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	kyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/crypto"
+	"github.com/openshift/service-ca-operator/pkg/controller/api"
+	"github.com/openshift/service-ca-operator/pkg/operator/util"
+)
+
+const (
+	signingCertificateLifetimeInDays = 365 // 1 year
+
+	// The minimum duration that the previous CA should be trusted after rotation is
+	// approximately half the default signing cert lifetime.
+	minimumTrustDuration = 180 * 24 * time.Hour
+)
+
+type unsupportedServiceCAConfig struct {
+	ForceRotation forceRotationConfig `json:"forceRotation"`
+}
+type forceRotationConfig struct {
+	Reason string `json:"reason"`
+}
+
+type signingCA struct {
+	config             *crypto.TLSCertificateConfig
+	bundle             []*x509.Certificate
+	intermediateCACert *x509.Certificate
+	oldCAExpiry        time.Time
+}
+
+// updateSigningSecret updates the provided secret with the signing artifacts.
+func (ca *signingCA) updateSigningSecret(secret *corev1.Secret) error {
+	caPEM, keyPEM, err := ca.config.GetPEMBytes()
+	if err != nil {
+		return err
+	}
+	bundlePEM, err := crypto.EncodeCertificates(ca.bundle...)
+	if err != nil {
+		return err
+	}
+	intermediatePEM, err := crypto.EncodeCertificates(ca.intermediateCACert)
+	if err != nil {
+		return err
+	}
+
+	d := secret.Data
+	d[corev1.TLSCertKey] = caPEM
+	d[corev1.TLSPrivateKeyKey] = keyPEM
+	d[api.BundleDataKey] = bundlePEM
+	d[api.IntermediateDataKey] = intermediatePEM
+
+	return nil
+}
+
+// rotateSigningSecret rotates the CA provided by the given secret and updates the secret
+// with the new CA's data.
```

Comment:
... or does nothing if neither a force rotation takes place nor the old ca expired.

Reply:
Done